From 4fc2215661fbcaa036fbdfe9e26aceb824405dce Mon Sep 17 00:00:00 2001
From: Fernando Guzman Lugo <fernando.lugo@ti.com>
Date: Wed, 15 Dec 2010 08:50:01 -0600
Subject: OMAP: mailbox- Empty work queues before destroying

Flush all the pending work items before destroying the
mbox queues. If not, it may cause a kernel panic if the
work_queue dereferences the mbox queue structure.

Change-Id: I84dc735b695cc4de72fea5e5830a4e58026a7c19
Signed-off-by: Fernando Guzman Lugo <fernando.lugo@ti.com>
---
 arch/arm/plat-omap/mailbox.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/arch/arm/plat-omap/mailbox.c b/arch/arm/plat-omap/mailbox.c
index dbba8a9..651e189 100644
--- a/arch/arm/plat-omap/mailbox.c
+++ b/arch/arm/plat-omap/mailbox.c
@@ -328,6 +328,10 @@ static void omap_mbox_fini(struct omap_mbox *mbox)
 	mutex_lock(&mboxes_lock);
 
 	if (!--mbox->use_count) {
+		omap_mbox_disable_irq(mbox, IRQ_RX);
+		omap_mbox_disable_irq(mbox, IRQ_TX);
+		flush_work(&mbox->rxq->work);
+		flush_work(&mbox->txq->work);
 		mbox_queue_free(mbox->txq);
 		mbox_queue_free(mbox->rxq);
 	}
-- 
1.6.0.4

